### ISSUE TITLE
LinuxMint pwsafe file corruption

### ISSUE THREAD

**Issue by pcismejr** [pwsafe/pwsafe#953]
I have a psafe3 file on a cloud server that I share with a LinuxMint PC, a Windows 11 desktop, and an Android phone. The password file was originally started on the Windows computer and is listed as a 3.14 file when viewed in properties. It has 644 entries. I have no issues with the Windows computer or the Android phone. I'm running 1.15 on the Linux machine. The file will open properly but if I make any changes, and save the file, most of the time the file is then corrupted. The Windows computer running 3.62 gives the error message that the file is corrupted, do I want to proceed, yes/no. The Linux computer will open the file but no longer shows the 'close' button and some entries are missing. I've narrowed the problem down to the PasswordSafe program on the Linux machine. I've eliminated the cloud server being any factor by testing the file locally on the Linux machine. I also tried reverting back to 1.12 but the issue is still present. Appreciate any help and willing to test further.

**Comment by pcismejr** [comment]
Thanks for the reply.

> Hi,
> 
> Are you using any non-English texts? PasswordSafe should fully support this, but that's my first suspect.

I'm using only English texts.

> 
> In any case, a sample database that reproduces the problem is the best way to debug this. Otherwise we're pretty much in the dark...

Not sure how to go about generating a sample database like the one I'm using. I might try exporting the entries to XML and then starting a new database using 1.15 in Linux and importing the XML. Then making changes and see if it corrupts. What's your opinion?

**Comment by ronys** [comment]
I was thinking of a sample database that doesn't have any sensitive data.

You might also try narrowing down the problem by deleting groups or entries and seeing if the problem replicates.

Alternately, export the file to XML on Windows, import it into a new file on Linux, as you suggested. If the problem still occurs, perhaps you can "sanitize" the XML file.

**Comment by gregdavisd** [comment]
I tried. Don't know what the cause is here. But I did find some suspicious code. In ItemField.cpp.

//Returns the number of bytes of 8 byte blocks needed to store 'size' bytes
size_t CItemField::GetBlockSize(size_t size) const
{
  return static_cast<size_t>(ceil(static_cast<double>(size) / 8.0)) * 8;
}

Using floating point maths to perform modular arithmetic. Maybe it works. But code can be compiled with an approximate math option. Some CPUs don't even have an 80 bit accurate FPU, everything is approximate. Which I shudder to think what would happen if 2.0 ended up equaling 1.99999999998.

I changed to: 

return ((size / 8) + ((size % 8 != 0 ? 1 : 0))) * 8;

As I believe the intention is to find an allocation size, rounded up to an 8 byte alignment.
